src/bindings/python/flux/hostlist.py

```diff
@@ -0,0 +1,208 @@
+        super().__init__()
+        self.pimpl = self.InnerWrapper(handle)
+
+    def __str__(self):
+        return self.encode()
+
+    def __repr__(self):
+        return f"Hostlist('{self}')"
+
+    def __len__(self):
+        return self.pimpl.count()
+
+    def __getitem__(self, index):
+        """Index and slice a hostlist
+
+        Works like normal Python list indexing, including slices:
+
+        >>> hl = Hostlist("foo[0-9]")
+        >>> hl[0]
+        'foo0'
+        >>> hl[9]
+        'foo9'
+        >>> hl[-1]
+        'foo9'
+        >>> hl[8:]
+        ['foo8', 'foo9']
+        >>> hl[1:3]
+        ['foo1', 'foo2']
+
+        """
+        if isinstance(index, numbers.Integral):
+            if index < 0:
+                index = len(self) + index
+            if 0 <= index < len(self):
+                # N.B. wrapper class already calls ffi.string() on result:
+                return self.pimpl.nth(index).decode("utf-8")
+            raise IndexError("Hostlist index out of range")
+
+        if isinstance(index, slice):
+            return [self[n] for n in range(len(self))[index]]
+
+        raise TypeError("Hostlist index must be integer or slice")
+
+    def __iter__(self):
+        """Return a Hostlist iterator"""
+        return HostlistIterator(self)
+
+    def __contains__(self, name):
+        """Test if a hostname is in a Hostlist"""
+        try:
+            self.pimpl.find(name)
+        except FileNotFoundError:
+            return False
+        return True
+
+    def encode(self):
+        """Encode a Hostlist to an RFC 29 hostlist string"""
+        #
+        #  N.B. Do not use automatic wrapper call here to avoid leaking
+        #   `char *` result. Instead explicitly call free() after copying
+        #   the returned string to Python
+        #
+        val = lib.hostlist_encode(self.handle)
+        result = ffi.string(val)
+        lib.free(val)
+        return result.decode("utf-8")
+
+    def count(self):
+        """Return the number of hosts in Hostlist"""
+        return self.pimpl.count()
+
+    def append(self, *args):
+        """Append one or more arguments to a Hostlist
+
+        Args may be either a Hostlist or any valid argument to Hostlist()
+        """
+        count = 0
+        for arg in args:
+            if not isinstance(arg, Hostlist):
+                arg = Hostlist(arg)
+            count += self.pimpl.append_list(arg)
+        return count
+
+    def delete(self, hosts):
+        """Delete host or hosts from Hostlist
+
+        param: hosts: A Hostlist or string in RFC 29 hostlist encoding
+        """
+        return self.pimpl.delete(str(hosts))
+
+    def sort(self):
+        """Sort a Hostlist"""
+        self.pimpl.sort()
+        return self
+
+    def uniq(self):
+        """Sort and remove duplicate hostnames from Hostlist"""
+        self.pimpl.uniq()
+        return self
+
+    def expand(self):
+        """Convert a Hostlist to a Python list"""
+        return self[:]
```

Comment:
You can potentially leave this method out.  Because `HostList` implements `__iter__`, users can just do `list(my_hostlist)`.